### PR TITLE
chore: asyncify downloading stuff

### PR DESCRIPTION
**Pull request by zsol**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #92
* #89
* __->__ #88
* #90

### CODE REVIEW

File: py_wtf/cli/__init__.py [hunk 1]
```diff
@@ -3,28 +3,35 @@
...
-    ProjectMetadata,
-    ProjectName,
-)
+from py_wtf.types import Documentation, FQName, Project, ProjectMetadata, ProjectName
+
+P = ParamSpec("P")
+T = TypeVar("T")
+
+
+def coroutine(f: Callable[P, Coroutine[None, None, T]]) -> Callable[P, T]:
```
Comment (abesto) [inline py_wtf/cli/__init__.py:29]: Oh no :D

File: py_wtf/indexer/pypi.py [hunk 2]
```diff
@@ -109,30 +111,43 @@ def parse_deps(maybe_deps: None | Sequence[str]) -> list[str]:
...
+
+        # this is a bit unnecessary 🙃
+        async with (
+            client.stream("GET", src_url) as response,
+            aiofiles.tempfile.NamedTemporaryFile("wb+", delete=False) as src_archive,
+        ):
+            async for chunk in response.aiter_bytes():
+                await src_archive.write(chunk)
+
+        with open_archive(src_archive.name) as opened:
```
Comment (abesto) [inline py_wtf/indexer/pypi.py:149]: Asyncify unpacking as well? Unsure if it makes a real difference in the end.

File: py_wtf/repository.py [hunk 3]
```diff
@@ -34,28 +40,36 @@ def _load_from_disk(self, key: ProjectName) -> None:
         index_file = self._index_file(key)
         index_contents = index_file.read_bytes()
         proj = converter.loads(index_contents, Project)
-        self._cache[key] = proj
+        loop = asyncio.get_event_loop_policy().get_event_loop()
+        self._cache[key] = Future(loop=loop)
+        self._cache[key].set_result(proj)
```
Comment (abesto) [inline py_wtf/repository.py:46]: I don't understand this part. We create a new future, and immediately set its result? (Same in `_save`). If we have the result, then why create a `Future` at all? If there was a `Future` in there already, then anyone `await`ing it will now block forever, because we'll never set the result of *that*.

File: py_wtf/cli/__init__.py (hunk 1 above)
Comment (zsol) [reply py_wtf/cli/__init__.py:29]: Yeah well this is what the `click` project recommends, so... :)

File: py_wtf/indexer/pypi.py (hunk 2 above)
Comment (zsol) [reply py_wtf/indexer/pypi.py:149]: No clue if it makes a difference but I didn't find any asyncio-aware tar+zip libraries. Worst case we can shove this in a threadpool executor

File: py_wtf/repository.py (hunk 3 above)
Comment (zsol) [reply py_wtf/repository.py:46]: This class needs a bit of rethinking, I just iterated until the tests passed :) It's a bit hard to follow

Comment (zsol) [reply py_wtf/repository.py:34]: It's been slightly rethunk.

File: py_wtf/repository.py [hunk 4]
```diff
@@ -34,28 +30,36 @@ def _load_from_disk(self, key: ProjectName) -> None:
...
+
+        # It's important to force the creation of this future before we await
+        # to make sure everyone awaits on the same future and so avoid duplicating work
+        fut = self._cache[key]
+
+        async for project in factory(key):
             self._save(project)
 
-        return self[key]
+        return await fut
```
Comment (zsol) [inline py_wtf/repository.py:65]: This is still a bit dodgy; if `factory(key)` never `async yield`s the `project` with `.name == key` then this `await` will block indefinitely (or until another poor coroutine happens to fulfil the future).

I'll fix this and add a test for it in a followup PR

File: py_wtf/repository.py [hunk 5]
```diff
@@ -13,19 +16,12 @@
 class ProjectRepository:
     directory: Path
 
-    _cache: dict[ProjectName, Project] = field(init=False)
+    _cache: dict[ProjectName, Future[Project]] = field(init=False)
 
     def __post_init__(self) -> None:
-        self._cache = {}
-
-    def __getitem__(self, key: ProjectName) -> Project:
```
Comment (abesto) [inline py_wtf/repository.py:21]: Nit: pass in `loop` to `ProjectRepository.__init__`?

File: py_wtf/repository.py (hunk 4 above)
Comment (abesto) [reply py_wtf/repository.py:65]: (y) At the very least on the happy path this will actually work like well-behaved futures!

**Review by abesto — APPROVED** [review]
